Best practices document (#25512)
sympy/sympy: doc/src/explanation/best-practices.md

# Best Practices

This page outlines some of the best practices for users of SymPy. The best
practices here will help avoid some common bugs and pitfalls that can occur
when using SymPy.

This page primarily focuses on best practices that apply generally to all
parts of SymPy. Best practices that are specific to certain SymPy submodules
or functions are outlined in the documentation for those specific functions.

## Basic Usage

(best-practices-defining-symbols)=
### Defining Symbols

- **The best way to define symbols is using the {func}`~.symbols` function.**

>   The `symbols()` function supports creating one or multiple symbols at a
>   time:

Reviewer: ```suggestion
  The `symbols()` function supports creating one or more symbols at once:
```

  ```py
  >>> from sympy import symbols
  >>> x = symbols('x')
  >>> a, b, c = symbols('a b c')
  ```

  Additionally, it supports adding assumptions to symbols

  ```py
  >>> i, j, k = symbols('i j k', integer=True)
  ```

  and defining {class}`~.Function` objects:

  ```py
  >>> from sympy import Function
  >>> f, g, h = symbols('f g h', cls=Function)
  ```

  It also supports shorthands for defining many numbered symbols at once:

  ```py
  >>> symbols('x:10')
  (x0, x1, x2, x3, x4, x5, x6, x7, x8, x9)
  ```

  There also exist some alternatives to {func}`~.symbols`. Two alternatives
  are using the {class}`~.Symbol` constructor directly and {mod}`sympy.abc`.
  These are both fine, but they are less general than {func}`~.symbols`, so

>   just using `symbols()` is generally preferred. {class}`~.Symbol` only

Reviewer 2: Is there a reason `symbols()` is preferred? I recall seeing the use of `Symbol()` in user code discouraged over the years, but never knew why. This doesn't really explain why.

Is it that we want to promote using `symbols()` in our documentation just for style sake? After reading this I'm guessing that all three methods should likely be equally promoted.

Reviewer 3: There is another reason why many people will not want to use symbols which is that it cannot be typed properly because it returns different types based on the value of its argument:
```python
In [1]: type(symbols('x'))
Out[1]: sympy.core.symbol.Symbol

In [2]: type(symbols('x y'))
Out[2]: tuple
```
The `cls=` argument could be typed properly but distinguishing strings like "x" vs "x y" is not something that any type system can do.

Also `Symbol` is more efficient if you just want to make one symbol.

This is sort of mentioned below "if you want to be sure you are only creating one symbol" but it might be worth clarifying why you might want to ensure this: for example if the symbol name is a string that is passed through some end user API then the user might pass a string with spaces e.g. `Matrix([[1,2],[3,4]]).charpoly('long name')`

Author: Symbol is fine. I just think it's simpler to have one function that can create one symbol or multiple symbols instead of one function that creates one symbol and another function that creates multiple symbols. With `symbols()`, you basically only ever need the one function, which simplifies things.

Reviewer 4: > you basically only ever need the one function, which simplifies things.

I just stopped using that in production code because `symbols` is unable to infer the types.
In the production code, you anyway are encouraged to do explicit imports from SymPy, so combined that you need to explicitly list you use in SymPy, I rather find that argument opinionated that that using one function having everything combined is really easier.

Reviewer 4: On the other hand, I'd like to listen to the opinion about if `symbols` is used inside for-loop sounds awkward or should be best practice.

Author: > Also if assumptions are involved, symbols doesn't make things easier because it sets assumptions simultaneously. And you eventually need to write one symbol for each lines to give according assumptions

In that case you end up with something like what is used in isympy. Different lines for each type of symbol. In my experience, most people just set everything to be real or positive anyway.

```py
>>> from sympy import *
>>> x, y, z, t = symbols('x y z t')
>>> k, m, n = symbols('k m n', integer=True)
>>> f, g, h = symbols('f g h', cls=Function)
```

> On the other hand, I'd like to listen to the opinion about if symbols is used inside for-loop sounds awkward or should be best practice.

I'm not following what you are asking here. What do you mean by using symbols in a for loop?

Reviewer 4: > I'm not following what you are asking here. What do you mean by using symbols in a for loop?

I would rather mean in other context, like building a web app where user can give any kind of equation. I don't use symbols because it parses the spacings.

I don't think that it's worth discussing whether to prefer symbols or symbol. 

In fact, the way how Symbol works behind is one of the easiest things that people can understand.
And it's better to let them trust their own sketches of how it works, than recommending something.

(I have other point of view that programming with type constructor as parameters like cls=Function is not becoming better nowadays because it limits type inference, but that's other topic)

Author: Before `symbols`, the common practice was to use `map(Symbol, 'xyz')`. But this is harder to read, and breaks down if you want symbol names with more than one character.

> In fact, the way how Symbol works behind is one of the easiest things that people can understand.

That's really not my experience, especially for people who are new to SymPy (and Python). The way symbols work in SymPy is surprising, especially for people coming from other systems where symbols are defined automatically. Even things like the fact that the variable name has no bearing on the symbol name throws a lot of people (people really have to work with Python for some time before they really *get* how variables work). 

I've seen people use things like `x = S('x')` to create a symbol, which is definitively **not** what you should be doing. Or people just parse full strings of expressions and try to do string expression manipulation (there are other sections in this guide about that). 

For these new users, `symbols` is the best choice.

I agree that if you have programmatic input, you should probably not be using `symbols`. That doesn't apply to 99% of SymPy users though.

Anyway, I'll probably rewrite this section a bit to be a little clearer on these points.

Author: I've rewritten this section.

  dictionary to [`parse_expr()`](sympy.parsing.sympy_parser.parse_expr). For example:

  **Don't**

  ```py
  >>> a, b, c = symbols('a b c', real=True)
  >>> # a, b, and c in expr are different symbols without assumptions
  >>> expr = parse_expr('a**2 + b - c')
  >>> expr.subs({a: 1, b: 1, c: 1}) # The substitution (apparently) doesn't work
  a**2 + b - c
  ```

  **Do**

  ```
  >>> # a, b, and c are the same as the a, b, c with real=True defined above
  >>> expr = parse_expr('a**2 + b - c', {'a': a, 'b': b, 'c': c})
  >>> expr.subs({a: 1, b: 1, c: 1})
  1
  ```

- Many operations on SymPy objects work as methods, not functions. These
  methods won't work on strings, since they are not yet SymPy objects. For
  example:

  ```
  >>> "x + 1".subs("x", "y")
  Traceback (most recent call last):
  ...
  AttributeError: 'str' object has no attribute 'subs'
  ```

  vs.

  ```
  >>> x, y = symbols('x y')
  >>> (x + 1).subs(x, y)
  y + 1
  ```

- Symbol names can contain any character, including things that aren't
  valid Python. But if you use strings as input, it is impossible to use such
  symbols. For example

  ```
  >>> from sympy import solve

>   >>> solve('x_{2} - 1') # doctest: +SKIP

Reviewer 4: I'm not sure if the LaTeX subscript is the best way to choose between `x_2, x_{2}, x2, ...` though
There is a problem that different expressions look same with same subscript, but compares differently `x_2, x_{2}, x2, ...`

Author: I agree that this is somewhat contrived. The point here is just to show how things can break, not to give a recommendation on how to name symbols. Anyway, I have seen users making symbol names like this.

>>> expression = x**2 + 0.5*x + 1
```

**Do**

```py
>>> from sympy import Rational
>>> expression = x**2 + Rational(1, 2)*x + 1
>>> expression = x**2 + x/2 + 1 # Equivalently
```

However, this isn't to say that you should never use floating-point numbers in
SymPy, only that if a more exact value is known it should be preferred. SymPy
does support [arbitrary precision floating-point
numbers](sympy.core.numbers.Float), but some operations may not perform as
well with them.

This also applies to non-rational numbers which can be represented exactly. For
example, one should avoid using `math.pi` and prefer `sympy.pi`, since the
former is a numerical approximation to $\pi$ and the latter is exactly $\pi$
(see also [](best-practices-separate-symbolic-and-numeric-code) below; in
general, one should avoid importing `math` when using SymPy).

**Don't**

```py
>>> import math
>>> import sympy
>>> math.pi
3.141592653589793
>>> sympy.sin(math.pi)
1.22464679914735e-16
```

**Do**

```py
>>> sympy.pi
pi
>>> sympy.sin(sympy.pi)
0
```

Here `sympy.sin(math.pi)` is not exactly 0, because `math.pi` is not exactly $\pi$.

One should also take care to avoid writing `integer/integer` where both

> integers are explicit integers. This is because Python will evaluate this to a floating-point value before SymPy is able to parse it.

Reviewer: ```suggestion
integers are explicit integers. This is because Python will evaluate
this to a floating-point value before SymPy is able to parse it.
```

Author: I'm having difficulty reading the GitHub diff. Is there an actual change to the text here or is this just a line wrap?

- A targeted simplification will not do something unexpected if the expression
  contains an unexpected form, or an unexpected subexpression. This is
  especially the case if simplification functions are applied with `deep=False`
  to only apply the simplification to the top-level expression.

Some other simplification functions are heuristical in nature, and care should
be taken with them as well. For example, the {func}`~.trigsimp` function is a
heuristic targeted to trigonometric functions, but the routines in the
{mod}`sympy.simplify.fu` submodule allow applying specific trigonometric
identities.

The [simplify section of the tutorial](tutorial-simplify) and the
[simplify module reference](../modules/simplify/simplify.rst) list the various
targeted simplification functions.

In some cases, you may know exactly what simplification operations you wish to
apply to an expression, but there may not be an exact set of simplification
functions that do them. When this happens, you can create your own targeted
simplification using {meth}`~sympy.core.basic.Basic.replace`, or in general, manually using
[advanced expression manipulation](tutorial-manipulation).

> <!-- TODO: Add an example -->

Author: Can anyone suggest a good example to demonstrate why using `simplify()` should be avoided (especially in programmatic contexts) over more targeted simplification functions like `cancel()`?

Reviewer: I think @oscarbenjamin mentioned one recently that had to do with hyperbolics appearing that made some roots more complicated. Trig functions without targeted (Fu-like) transpormation are notorious for getting unwieldy. Also, even something simple like 
```python
>>> exp(x)/2 + exp(-x)/2
exp(x)/2 + exp(-x)/2
>>> simplify(_)
cosh(x)
```
is simpler but one might not have wanted to convert to hyperbolics.

Reviewer 3: The example from https://github.com/sympy/sympy/pull/25458#issuecomment-1662397968 is a good one:
```python
In [10]: M = Matrix([[0, -t*cos(t) + sin(t)], [-t*cos(t) + sin(t), 0]])

In [11]: M
Out[11]: 
⎡        0           -t⋅cos(t) + sin(t)⎤
⎢                                      ⎥
⎣-t⋅cos(t) + sin(t)          0         ⎦

In [12]: p = M.charpoly(z).as_expr()

In [13]: p
Out[13]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [14]: roots(p, z)
Out[14]: 
⎧    ___________________________________        ___________________________________   ⎫
⎨   ╱  2    2                      2           ╱  2    2                      2       ⎬
⎩-╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1, ╲╱  t ⋅cos (t) - t⋅sin(2⋅t) + sin (t) : 1⎭
```
Why didn't roots spot that there is a simpler expression for the roots here? Where did `sin(2*t)` come from?

The answer is:
```python
In [15]: p.expand()
Out[15]: 
   2    2                           2      2   
- t ⋅cos (t) + 2⋅t⋅sin(t)⋅cos(t) + z  - sin (t)

In [16]: p.expand().simplify()
Out[16]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)

In [17]: p.expand().collect(z, factor)
Out[17]: 
 2                      2
z  - (t⋅cos(t) - sin(t)) 

In [18]: p.expand().simplify().collect(z, factor)
Out[18]: 
   2    2                    2      2   
- t ⋅cos (t) + t⋅sin(2⋅t) + z  - sin (t)
```
So somewhere `simplify` (or `trigsimp`) is used and it decides to replace `2*sin(t)*cos(t)` with `sin(2*t)`. Whether or not this is a good idea is very context dependent but if you intend to do algebra like `factor`, `cancel` etc then it is definitely not better to introduce new "generators" (although eliminating some generators in favour of others that are *already* present is usually good). Here the use of `simplify` turns an expression that could have been factorised into one that cannot be factorised. In turn this takes us from nice rational expressions for the roots that are free of branch cuts to horrible trig-radical expressions.

I'm not sure why `eigenvals` itself is not affected but all of these functions misuse or overuse `simplify`:
```python
In [20]: M.eigenvals()
Out[20]: {-t⋅cos(t) + sin(t): 1, t⋅cos(t) - sin(t): 1}
```
The fact that `simplify` is actually being used *inside* functions like `roots`, `eigenvals` etc means that it is not even possible to carefully manipulate your inputs into the form you want before calling those functions.

Author: I'll have to think about this. A potential problem with an example like this is that it's hard to not just say "well, shouldn't `simplify()` just be smarter?" (e.g., in this case, try to rewrite the expression in such a way that makes it more factorable before calling `factor()`). And indeed, it's quite possible that at some point in the future it *will* become smarter, which would completely break the example.

It would be more robust to have an example where simplify() pretty unambiguously returns a "simple" form of an expression by some measure, but you can also say by some other measure that it isn't simple.

Maybe I can manipulate this example into something like that, but it's hard because you're really just saying "`simplify()` returned an expression that wasn't the correct 'simplified' for `simplify()`". But it would be I think more instructive to show an example like "`simplify()` returned something that wasn't the correct 'simplified' for `<some other function>`".

Another idea: even just for a basic rational function, there's at least 3 "simple" ways it can be represented: as p/q with p and q expanded and no factors in common (i.e., `cancel()`), with p/q with p and q factored (`factor()`), or as a partial fraction decomposition (`apart()`). `simplify()` seems to prefer the first of these, but there's no reason in particular for that, and it might some day switch to one of the others. But the point is that there are situations where all three forms can be useful.

Reviewer 3: I think that anything that you might write here would be potentially subject to change in future because simplify is heuristic. The point of my example is that it is a real problem that is actually caused by the use of simplify right now in the SymPy codebase. Of course we *should* try to fix that in future but it demonstrates why naive use of simplify is not always good (besides speed concerns) because it can lead to unpredictable results that ruin a larger calculation.

Reviewer 4: I think that the problem of simplify

- Ambiguity problem:
  In the PR https://github.com/sympy/sympy/pull/25456, it is discovered that simplify is very hard to trigger and is very difficult to quantify when does it trigger.

- Abstraction problem:
  In the dependency inversion principle, high level features shouldn't depend on the low level implementation, but abstractions of it.
 `simplify` is very specific to the implementation details of computer algebra itself, than the abstract algebra. For example, I'm very skeptical about if transfer function should compute feedback result, or quaternions should compute norms with `simplify` than do bare algebra.
  This can later cause significant restriction if you want to let users of use implementations of symengine, or do other customizations.

Reviewer 4: I think that the only production and practical usage of `simplify` should be for compilation workflow.
And similar as how compiler works, the compiled code can never be reused by humans, and can only interpreted by compilers or machines.
Similar as that, the result of simplify is never interpretable than other `simplify` or the printers.

Reviewer 3: I don't think that `simplify` should even be used for a compilation workflow: if you want to optimise an expression for turning into compiled code then you want a specialised simplify function for that.

Basically I don't think there is any situation within SymPy itself where `simplify` should be used but it is a nice thing for users to use interactively that will likely do what they want in simple cases.

Author: That's actually a good point. I think in practice, more people have the issue that simplify does too little, that it doesn't happen to actually call the simplification function they are looking for.

  ```py
  if x > 0:
      expr = 1
  else:
      expr = 0
  ```

  this can be represented symbolically as

  ```py
  >>> from sympy import Piecewise, pprint
  >>> expr = Piecewise((1, x > 0), (0, True))
  >>> pprint(expr, use_unicode=True)
  ⎧1  for x > 0
  ⎨
  ⎩0  otherwise
  >>> expr.subs(x, 1)
  1
  >>> expr.subs(x, -1)
  0
  ```

- **Use {func}`~.ordered` to sort expressions into a canonical order.** If you
  are trying to use `sorted` because you want a canonical ordering, but you
  don't particularly care what that ordering is, you can use `ordered`.

  ```py
  >>> from sympy import ordered
  >>> list(ordered([x, 0]))
  [0, x]
  ```

  Alternatively, try to write the function in a way so that the correctness of
  the result does not depend on the order that arguments are processed in.

## Custom SymPy Objects

SymPy is designed to be extended with custom classes, typically by subclassing
{term}`Basic`, {term}`Expr`, or {term}`Function <Function (class)>`. All the
symbolic classes in SymPy itself are written this way, and the points here
apply equally to them as to user-defined classes.

For an in-depth guide on how to write a `Function` subclass, see the [guide on

> writing custom functions](custom-functions).

Reviewer 4: My opinion is SymPy rarely need to be extended though.

For example, I'm not sure about what kind of things are more capable to do with custom functions with `eval`, than `Function` with substitution)

Author: Both this guide and the one linked here apply equally well to user code and library code. Every class in SymPy is a "custom object".

Author: Basically, I wrote this section of this guide and the custom functions guide primarily for for SymPy developers, because we were sorely lacking in that sort of documentation and there are a lot of things involved. But everything there is public API and completely supported for users. It's true that the vast majority of users shouldn't be doing anything this advanced, but there are some who do, and there are legitimate reasons to do so. We should always aim to make the methods used to extend SymPy internally to also be usable by end-users. That way we keep the APIs for doing so clean and sufficiently extensible.

  boolean flag). Code duplication can be avoided by using a common superclass.

- **If the data you want to store is a Python function**, it's best to just
  use as a method on the class. In many cases, the method may already fit into
  one of the [existing set of overridable SymPy methods](custom-functions). If
  you want to define how a function evaluates itself numerically, you can use
  {func}`~.implemented_function`.

- **Represent the information using by modifying the object's `func`.** This
  solution is much more complicated than the others, and should only be used
  when it is necessary. In some extreme cases, it is not possible to represent
  every mathematical aspect of an object using `args` alone. This can happen,
  for example, because of the limitation that [`args` should only contain
  `Basic` instances](best-practices-args-invariants). It is still possible to
  create custom SymPy objects in these situations by using a custom
  {term}`func` that is different from `type(expr)` (in this case, you would
  override `__eq__` on the `func` [rather than on the
  class](best-practices-eq)).

  However, this sort of situation is rare.

(best-practices-eq)=

> ### Don't Overwrite `__eq__`

Reviewer 3: I don't know if we want to mention this here but related to "avoid extra attributes" and "don't override `__eq__`" is the question of what to do when you *do* need to do these things.

It is rare that these things are needed but there are examples e.g. `Poly`. Storing the `args` of a Poly as Expr would be very inefficient so it is not done and also the domain argument is not representable as `Basic` (possibly what this shows is that `Poly` should not be `Basic` at all). Likewise `ImmutableMatrix` has a `DomainMatrix` as its internal representation and it would be inefficient to always store the `args` in that case.

A more common situation where you do need to attach additional data not in the `args` is for atoms.

The question then is what you *should* do e.g. to implement an atom or some other Basic subclass whose constructor arguments will *not* end up in the args or at least will not be stored explicitly in `_args`.

1. Don't override `__eq__` or `__hash__`: use `_hashable_content` instead.
2. Override `__new__` but *don't* pass the args to `super().__new__`.
3. Implement `__getnewargs__` (`__getnewargs_ex__` if keyword arguments are needed) for pickling etc.
4. If not an `Atom` then override `.args` as a property that will generate args on demand (and ensure that the `func-args` invariant holds).

Author: These are somewhat advanced things. I'm not even sure myself what the proper way to do some of those things would be (is it ever actually a good idea to override `_hashable_content`?). In my experience, most people who are trying to override `__eq__` are completely misguided and shouldn't be doing it at all. They are just trying to do something that goes against the SymPy norms.

I did mention overriding `func` as an advanced escape hatch. Maybe I can briefly mention these other things, but if anything, I feel like this just gets to the core of why this is a bad idea: there are lot of things here and it's very easy to get it wrong.

  control by defining a `_sympy_` method and for objects you do not control by
  extending the `converter` dictionary. See the {func}`~.sympify`
  documentation for more details.

### Avoiding Infinite Recursion from Assumptions Handlers

When writing assumptions handlers on custom functions like `_eval_is_positive`
(see the [custom
functions guide](custom-functions-assumptions) for details on how to do this),
there are two important things to keep in mind:

**Firstly, avoid creating new expressions inside of an assumption handler. You
should always pull apart the arguments of a function directly instead.** The
reason is that creating a new expression could itself result in an assumptions
query. This can easily lead to infinite recursion. And even when it doesn't,
creating a new expression that itself could lead to many recursive
assumptions queries is bad for performance compared to querying the desired
property more directly.

This generally means using methods like {meth}`~.as_independent` or
`{meth}`~.as_coeff_mul` and checking
the `args` of expressions directly (see the [custom functions
guide](custom-functions-assumptions) for an example).

> <!-- TODO: Give an example here. Can we show an infinite recursion? -->

Author: @oscarbenjamin do you know of an example of when assumptions handles can actually lead to infinite recursion? I couldn't come up with one easily, but I think maybe the problem is that this mainly happens inside of core SymPy classes rather than custom classes because those are the ones that are more likely to call each other.

Reviewer 3: > do you know of an example of when assumptions handles can actually lead to infinite recursion?

You have an example below. Do you mean something from the codebase though?

Generally I would remove any code that does this from the codebase.

Author: I mean an illustrative example I can include here. The example below is based on infinite recursion from evaluating assumptions on `self`, but I am looking for an example based on the "don't create new expressions in assumptions handlers" rule.

Reviewer 3: I can contrive an example but as you say it mostly happens in core class e.g. because someone does not consider that `1 + self` creates a new `Add` or something like that.

Here are examples of what should not be done:
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L618
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L661-L664
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L680
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L746
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L748-L750
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L754-L756

There are many more in the same file. Maybe you can identify an immediate bug from those. There are two types of infinite recursion that can occur:

- Recursing onto a different object that then recurses onto another different object and so on.
- Recursing onto a larger object that contains the current object so that the query comes back down to the same query on the original object.

It is very difficult to predict when any created expressions might lead to either of these cases because automatic evaluation rules can be so unpredictable. I don't see how else to define a rule that could control it besides "never create a new expression".

Author: To be fair, some of those seem hard to do without just completely copying the assumptions logic from the other class. I guess I basically already said the same thing in the other issue, but this really is a fundamental limitation of this recursive way of implementing assumptions handlers, which makes them responsible for both returning the result of the assumptions query and, to some extent, the method by which other queries are logically deduced. A better design is to separate out the order that the queries are actually made in to a top-level solver, and represent facts in an symbolic way.

Anyway, I think this is probably hard to show an explicit example for. Maybe there's an issue or PR I could reference where something like this happened. Otherwise, I'm just going to leave this as it is and remove the TODO.
